src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocBlockTagLocationCheck.java

```diff
@@ -0,0 +1,199 @@
+ * to the middle of a line, it will be treated as plain text without any
+ * violations, but the generated documentation will be erroneous.
+ * </p>
+ * <p>
+ * To place a tag explicitly as text, escape the {@code @} symbol with &amp;#64;
+ * or place it inside {@code {@code }}, for example:
+ * </p>
+ * <pre>
+ * &#47;**
+ *  * &amp;#64;serial literal in {&#64;code &#64;serial} Javadoc tag.
+ *  *&#47;
+ * </pre>
+ * <ul>
+ * <li>
+ * Property {@code tags} - Specify the javadoc tags to process.
+ * Default value is {@code author, deprecated, exception, hidden, param, provides,
+ * return, see, serial, serialData, serialField, since, throws, uses, version}.
+ * </li>
+ * <li>
+ * Property {@code violateExecutionOnNonTightHtml} - Control when to print violations
+ * if the Javadoc being examined by this check violates the tight html rules defined at
+ * <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML Rules</a>.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the default check:
+ * </p>
+ * <pre>
+ * &lt;module name="JavadocBlockTagLocation"/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * &#47;**
+ *  * Escaped tag &amp;#64;version (OK)
+ *  * Plain text with {&#64;code &#64;see} (OK)
+ *  * A @custom tag (OK)
+ *  * &#64;since 1.0 (OK)
+ *  * text &#64;return (violation)
+ *  * * &#64;param (violation)
+ * +* &#64;serial (violation)
+ *  * &#64;see first (OK) &#64;see second (violation)
+ *  *&#47;
+ * public int field;
+ * </pre>
+ * <p>
+ * To configure the check to verify tags from
+ * <a href="https://openjdk.java.net/jeps/8068562">JEP 8068562</a> only:
+ * </p>
+ * <pre>
+ * &lt;module name="JavadocBlockTagLocation"&gt;
+ *   &lt;property name="tags" value="apiNote, implSpec, implNote"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>
+ * To configure the check to verify all default tags and some custom tags in addition:
+ * </p>
+ * <pre>
+ * &lt;module name="JavadocBlockTagLocation"&gt;
+ *   &lt;!-- default tags --&gt;
+ *   &lt;property name="tags" value="author, deprecated, exception, hidden"/&gt;
+ *   &lt;property name="tags" value="param, provides, return, see, serial"/&gt;
+ *   &lt;property name="tags" value="serialData, serialField, since, throws"/&gt;
+ *   &lt;property name="tags" value="uses, version"/&gt;
+ *   &lt;!-- additional tags used in the project --&gt;
+ *   &lt;property name="tags" value="noinspection"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ *
+ * @since 8.24
+ */
+@StatelessCheck
+public class JavadocBlockTagLocationCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties" file.
+     */
+    public static final String MSG_BLOCK_TAG_LOCATION = "javadoc.blockTagLocation";
+
+    /**
+     * This regexp is used to extract the javadoc tags.
+     */
+    private static final Pattern JAVADOC_BLOCK_TAG_PATTERN = Pattern.compile("\\s@(\\w+)");
```

Comment:
Please explain what is missing in AST that force you to use extra parsing.

Reply:
The parser is fine. It does exactly what [the specs](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javadoc.html#javadoctags) say:

> A tag must start at the beginning of a line (after any leading spaces and an optional asterisk) or it is treated as normal text:

```
    |   |   |   |       |--JAVADOC_TAG -> JAVADOC_TAG [11:7]
    |   |   |   |       |   |--PARAM_LITERAL -> @param [11:7]
    |   |   |   |       |   |--WS ->   [11:13]
    |   |   |   |       |   |--PARAMETER_NAME -> param1 [11:14]
    |   |   |   |       |   |--WS ->   [11:20]
    |   |   |   |       |   `--DESCRIPTION -> DESCRIPTION [11:21]
    |   |   |   |       |       |--TEXT -> a parameter @param, @custom //warn [11:21]
    |   |   |   |       |       `--NEWLINE -> \n [11:55]
```

Reply:
we need to docuemnt this , it is not obvious  ...See comment below